urlaa914a1de2_xu_hao_erilex_tgz-pJ8-erilex_esl_ESLJ8/src/erilex/codegen/embedded/TransformerLanguageGenerator.java

```diff
@@ -0,0 +1,558 @@
+        }
+    }
+
+    public static String getNodeEnvStr(NonTerminal nt) {
+        if (nt.envType != null) {
+            return nt.envType.unparse();
+
+        } else {
+            return nt.env;
+        }
+    }
+
+    public void types(NameEmbedding t, LanguageBuilder T, String ss) {
+        t.add("Eta", Generator.Eta.cons, true, false);
+        t.add(dnchi.cons, dchi.cons, true, true);
+        t.add(ntchi.cons, ntchi.cons, true, false);
+        t.add(dchi.cons, dchi.cons, true, false);
+        t.add(dchi1.cons, dchi1.cons + "1", true, false);
+        t.add(dchi2.cons, dchi2.cons + "2", true, false);
+        t.add("_Class", "java.lang.Class", false, true); // TODO change this to ? extends
+        t.add("Boolean", "java.lang.Boolean", false, true);
+        // This is used to ensure that type F has an APP method.
+        // A class F without any method is first generated
+        // and then overriden by the utility class.
+        t.add("F", "F", false, false);
+
+        T.def(ss).start()
+                .t("F").nt(ss).nt(ss)
+                .or().t("V").nt(ss)
+                .or().t("List").nt(ss)
+                .or().t(TRL_NODE_CONS).nt(NODE_TYPE).nt(ss).end();
+        T.def(NODE_TYPE).start()
+                .t("ntChi").end();
+    }
+    private int anyn = 0;
+    private Type getProductionTypeStr(NameEmbedding Gt, NameEmbedding Gd, Grammar GT, Production p) {
+        String anynstr = ANY+anyn++;
+        T.def(T.startSymbol).start().t(anynstr).end();
+        t.add(anynstr, "_"+anynstr, true, false);
+        Type ty = getAbstactProductionOriginalTypeStr((AbstractProduction) p,GT, Gt,Generator.tvar(anynstr));
+        if (Gd.nat(p.getProdName())) {
+            return ty;
+        } else {
+            return cons(TRL_NODE_CONS,cons(OL_PREFIX+p.getProdName()), ty);
+        }
+    }
+
+    public static class Eval {
+        Grammar OLG;
+        HashMap<String, Pair<String, Integer>[]> parameterMap = new HashMap();
+
+        public Eval(Grammar OLG) {
+            this.OLG = OLG;
+            ArrayList<Production> seqList =new ArrayList<Production>();
+            ArrayList<Production> elemList =new ArrayList<Production>();
+            for(Production p : OLG.ruleMap.values()) {
+                String name = p.getProdName();
+                seqList.clear();
+                LanguageGenerator.breakOrList(p, seqList);
+                Pair<String, Integer>[] parameters = new Pair[seqList.size()];
+                int i = 0;
+                for(Production p2 : seqList) {
+                    elemList.clear();
+                    LanguageGenerator.breakSeqList(p2, elemList);
+                    parameters[i++] = new Pair(((Terminal)elemList.get(0)).text, elemList.size() - 1);
+                }
+                    parameterMap.put(name, parameters);
+            }
+        }
+        public Tree<ASTValueData> eval(Pair env, Tree<ASTValueData> expr) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 20. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.